server.js

```diff
@@ -2,32 +2,162 @@ import express from 'express'
+      return (
+        item.authors.toLowerCase().replace("-", " ").split(" ").includes(author.toLowerCase()) || // will match against one name in author name
+        item.authors.toLowerCase().split("-").includes(author.toLowerCase()) // will match against full author name
+        // could use instead: RegExp("\\b" + author.toLowerCase() + "\\b").test(item.authors.toLowerCase())
+      )
+    })
+    if (filteredBooksOnAuthor.length > 0) {
+      return response.json(filteredBooksOnAuthor.slice(0, 20))
+    } else if (filteredBooksOnAuthor.length === 0) {
+      return response.status(404).json({ error: "No such author" })
+    }
+  }
+
+  else if (language && !author) {
+    const filteredBooksOnLanguage = booksData.filter(item => {
+      return (item.language_code === language)
+    })
+    if (filteredBooksOnLanguage.length > 0) {
+      return response.json(filteredBooksOnLanguage.slice(0, 20))
+    } else if (filteredBooksOnLanguage.length === 0) {
+      return response.status(404).json({ error: "What is that language?" })
+    }
+  }
+
+  else if (author && language) {
+    const filteredBooksOnAuthorAndLanguage = booksData.filter(item => {
+      return (
+        (item.authors.toLowerCase().replace("-", " ").split(" ").includes(author.toLowerCase()) // will match against one name in author name
+          ||
+          item.authors.toLowerCase().split("-").includes(author.toLowerCase()))  // will match against full author name
+        // could use instead: RegExp("\\b" + author.toLowerCase() + "\\b").test(item.authors.toLowerCase())
+        &&
+        item.language_code === language
+      )
+    })
+    if (filteredBooksOnAuthorAndLanguage.length > 0) {
+      return response.json(filteredBooksOnAuthorAndLanguage.slice(0, 20))
+    } else {
+      return response.status(404).json({ error: "No such combination" })
+    }
+  }
+
+
+  else if (toprated) {
+    if (+toprated > booksData.length) {
+      return response.status(404).json({ error: "That is just too many books" })
+    } else {
+      const sortedBooksOnRating = booksData.sort(function (a, b) {
+        return b.average_rating - a.average_rating
+      })
+      const topratedBooks = sortedBooksOnRating.slice(0, toprated)
+      return response.json(topratedBooks)
+    }
+  }
+
+  else if (shortest) {
+    const minPages = 20 // business rule: pages less than 20 is an incorrect value
+    const filteredBooksOnPages = booksData.filter(item => {
+      return (item.num_pages >= minPages)
+    })
+    if (+shortest > filteredBooksOnPages.length) {
+      return response.status(404).json({ error: "That is just too many books" })
+    } else {
+      const sortedBooksOnPages = filteredBooksOnPages.sort(function (a, b) {
+        return a.num_pages - b.num_pages
+      })
+      const shortestBooks = sortedBooksOnPages.slice(0, shortest)
+      return response.json(shortestBooks)
+    }
+  }
+
+  else {
+    return response.json(booksData.slice(0, 20))
+  }
```

Comment:
Instead of having all these if/else, what do you think about refiltering the same data?

```
let filteredBooks
if (author) {
  filteredBooks  = //filtered on author
}
if (language) {
  filteredBooks = //filtered on language
}
```
and so on? :)